File: s3-plugin-server/src/main/java/com/gu/teamcity/S3Extension.java
package com.gu.teamcity;

import jetbrains.buildServer.serverSide.MainConfigProcessor;
import jetbrains.buildServer.serverSide.SBuildServer;
import org.jdom.Element;

public class S3Extension implements MainConfigProcessor {
    public static final String bucketElement = "bucketElement";
    public static final String s3Element = "S3";

    public String bucketName;

    public S3Extension(SBuildServer server) {
        server.registerExtension(MainConfigProcessor.class, "S3", this);
    }

    public void readFrom(Element root) {
        Element s3 = root.getChild(s3Element);
        if (s3 != null && s3.getChild(bucketElement) != null) {
            bucketName = root.getAttributeValue("name");
        }
    }

    public void writeTo(Element root) {
        if (bucketName != null) {
            Element s3 = new Element(s3Element);
            Element bucket = new Element(bucketElement);
            bucket.setAttribute(bucketElement, bucketName);
            s3.addContent(bucket);
            root.addContent(s3);
        }
    }
}

File: s3-plugin-server/src/main/scala/com/gu/teamcity/S3Extension.scala
package com.gu.teamcity

import jetbrains.buildServer.serverSide.MainConfigProcessor
import jetbrains.buildServer.serverSide.SBuildServer
import org.jdom.Element

class S3Extension extends MainConfigProcessor {
  var bucketName: Option[String] = None

  def readFrom(root: Element) {
    for {
      s3 <- Option(root.getChild(S3Extension.s3Element))
      bucket <- Option(s3.getChild(S3Extension.bucketElement))
    }
      bucketName = Option(root.getAttributeValue("name"))
  }

  def writeTo(root: Element) {
    for (name <- bucketName) {
      val s3 = new Element(S3Extension.s3Element)
      val bucket = new Element(S3Extension.bucketElement)
      bucket.setAttribute(S3Extension.bucketElement, name)
      s3.addContent(bucket)
      root.addContent(s3)
    }
  }
}

object S3Extension {
  val bucketElement = "bucket"
  val s3Element = "S3"
}